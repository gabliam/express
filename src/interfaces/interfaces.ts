import * as express from 'express';
import { interfaces } from '@gabliam/core';


export type HandlerDecorator = (target: any, key: string, value: any) => void;

export interface ExpressPluginConfig {
    rootPath: string;

    port: number;

    hostname: string;
}

export type MiddlewareConfigurator = (...values: any[]) =>  express.RequestHandler |Â express.RequestHandler[];

export interface MiddlewareDefinition {
    name: string;

    values: any[];
}

/**
 * Config function
 */
export type ConfigFunction = (app: express.Application) => void;

export interface ExpressConfigRegistry extends interfaces.ValueRegistry {
    key: string;
}
