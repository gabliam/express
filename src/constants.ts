/** Type in registry */
export const TYPE = {
  Controller: 'ControllerType'
};

export const METADATA_KEY = {
  controller: '_controller',
  controllerMethod: '_controller-method',
  controllerParameter: '_controller-parameter',
  middleware: '_middleware',
  MiddlewareConfig: '_middlewareConfig',
  MiddlewareErrorConfig: '_middlewareErrorConfig'
};

export enum PARAMETER_TYPE {
  REQUEST,
  RESPONSE,
  PARAMS,
  QUERY,
  BODY,
  HEADERS,
  COOKIES,
  NEXT
}

export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const EXPRESS_PLUGIN_CONFIG = Symbol('GABLIAM/EXPRESS_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/EXPRESS_APP');
export const SERVER = Symbol('GABLIAM/EXPRESS_SERVER');
