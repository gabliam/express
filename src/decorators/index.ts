export * from './controller';
export * from './middleware';
export * from './config';
export * from './method';
export * from './params';
