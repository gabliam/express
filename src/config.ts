import { PluginConfig, Bean, Value } from '@gabliam/core';
import { DEFAULT_ROUTING_ROOT_PATH, EXPRESS_PLUGIN_CONFIG } from './constants';
import { ExpressPluginConfig } from './interfaces';
import * as Joi from 'joi';

@PluginConfig()
export class RestPluginConfig {
    @Value('application.express.rootPath', Joi.string())
    rootPath = DEFAULT_ROUTING_ROOT_PATH;

    @Value('application.express.port', Joi.number().positive())
    port: number = process.env.PORT || 3000;

    @Value('application.express.hostname', Joi.string())
    hostname: string;


    @Bean(EXPRESS_PLUGIN_CONFIG)
    restConfig(): ExpressPluginConfig {
        return {
            rootPath: this.rootPath,
            port: this.port,
            hostname: this.hostname
        };
    }
}
