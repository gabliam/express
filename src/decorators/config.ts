import { METADATA_KEY } from '../constants';
import { ExpressConfigMetadata } from '../interfaces';

export function ExpressConfig() {
  return ExpressConfigDecorator(METADATA_KEY.MiddlewareConfig);
};

export function ExpressErrorConfig() {
  return ExpressConfigDecorator(METADATA_KEY.MiddlewareErrorConfig);
};

function ExpressConfigDecorator(type: string) {
  return function (target: any, key: string, descriptor: PropertyDescriptor) {
    const id = target.constructor.name;
    const metadata: ExpressConfigMetadata = { id, key };
    let metadataList: ExpressConfigMetadata[] = [];

    if (!Reflect.hasOwnMetadata(type, target.constructor)) {
      Reflect.defineMetadata(type, metadataList, target.constructor);
    } else {
      metadataList = Reflect.getOwnMetadata(type, target.constructor);
    }

    metadataList.push(metadata);
  };
};
