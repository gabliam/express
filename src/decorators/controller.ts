import * as express from 'express';
import { ControllerMetadata } from '../interfaces';
import { METADATA_KEY, TYPE, ERRORS_MSGS } from '../constants';
import { inversifyInterfaces, injectable, register } from '@gabliam/core';
import { addMiddlewareMetadata } from '../metadata';

export interface ControllerOptions {
  name?: string;

  path: string;

  middlewares?: express.RequestHandler[];
}

export function Controller(options: ControllerOptions | string) {
  return function(target: any) {
    decorateController(options, target, false);
  };
}

export function RestController(options: ControllerOptions | string) {
  return function(target: any) {
    decorateController(options, target, true);
  };
}

function decorateController(
  options: ControllerOptions | string,
  target: any,
  json: boolean
) {
  if (Reflect.hasOwnMetadata(METADATA_KEY.controller, target) === true) {
    throw new Error(ERRORS_MSGS.DUPLICATED_CONFIG_DECORATOR);
  }

  let path: string;
  let id: inversifyInterfaces.ServiceIdentifier<any> = target;
  let middlewares: express.RequestHandler[] = [];
  if (typeof options === 'string') {
    path = options;
  } else {
    path = options.path;
    middlewares = options.middlewares || [];
    if (options.name) {
      id = options.name;
    }
  }

  addMiddlewareMetadata(middlewares, target);

  const metadata: ControllerMetadata = { path, json };
  Reflect.defineMetadata(METADATA_KEY.controller, metadata, target);
  injectable()(target);
  register(TYPE.Controller, { id, target })(target);
}
