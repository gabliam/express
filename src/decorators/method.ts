import {
  HandlerDecorator,
  ControllerMethodMetadata,
  MiddlewareMetadata
} from '../interfaces';
import { METADATA_KEY } from '../constants';
import { addMiddlewareMetadata } from '../metadata';

export function All(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('all', path, ...middlewares);
}

export function Get(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('get', path, ...middlewares);
}

export function Post(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('post', path, ...middlewares);
}

export function Put(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('put', path, ...middlewares);
}

export function Patch(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('patch', path, ...middlewares);
}

export function Head(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('head', path, ...middlewares);
}

export function Delete(
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return Method('delete', path, ...middlewares);
}

export function Method(
  method: string,
  path: string,
  ...middlewares: MiddlewareMetadata[]
): HandlerDecorator {
  return function(target: any, key: string, descriptor: PropertyDescriptor) {
    const metadata: ControllerMethodMetadata = { path, method, key };
    let metadataList: ControllerMethodMetadata[] = [];

    addMiddlewareMetadata(middlewares, target.constructor, key);
    if (
      !Reflect.hasOwnMetadata(METADATA_KEY.controllerMethod, target.constructor)
    ) {
      Reflect.defineMetadata(
        METADATA_KEY.controllerMethod,
        metadataList,
        target.constructor
      );
    } else {
      metadataList = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        target.constructor
      );
    }

    metadataList.push(metadata);
  };
}
