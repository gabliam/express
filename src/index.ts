import { interfaces as coreInterfaces, inversifyInterfaces, Scan, Registry } from '@gabliam/core';
import { TYPE, METADATA_KEY, EXPRESS_PLUGIN_CONFIG, APP, SERVER } from './constants';
import { getMiddlewares } from './metadata';
import { cleanPath } from './utils';
import * as interfaces from './interfaces';
import * as express from 'express';
import * as d from 'debug';
import * as http from 'http';

const debug = d('Gabliam:Plugin:ExpressPlugin');

export * from './decorators';
export { interfaces, APP, SERVER };

@Scan(__dirname)
export default class ExpressPlugin implements coreInterfaces.GabliamPlugin {
  middlewares: interfaces.ExpressConfig[] = [];

  errorMiddlewares: interfaces.ExpressConfig[] = [];

  bind(container: inversifyInterfaces.Container, registry: Registry) {
    container.bind(APP).toConstantValue(express());
    registry.get(TYPE.Controller)
      .forEach(({ id, target }) => container.bind<any>(id).to(target).inSingletonScope());
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    this.buildExpressConfig(container, registry);
    this.buildControllers(container, registry);
    this.buildExpressErrorConfig(container, registry);
  }

  config(container: inversifyInterfaces.Container, registry: Registry, confInstance: any) {
    if (Reflect.hasMetadata(METADATA_KEY.MiddlewareConfig, confInstance.constructor)) {
      const metadataList: interfaces.ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        this.middlewares.push({ order, instance: confInstance[key].bind(confInstance[key]) })
      });
    }

    if (Reflect.hasMetadata(METADATA_KEY.MiddlewareErrorConfig, confInstance.constructor)) {
      const metadataList: interfaces.ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        this.errorMiddlewares.push({ order, instance: confInstance[key].bind(confInstance[key]) });
      });
    }
  }

  private buildExpressConfig(container: inversifyInterfaces.Container, registry: Registry) {
    const app = container.get<express.Application>(APP);
    this.middlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildExpressErrorConfig(container: inversifyInterfaces.Container, registry: Registry) {
    const app = container.get<express.Application>(APP);
    this.errorMiddlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildControllers(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<interfaces.ExpressPluginConfig>(EXPRESS_PLUGIN_CONFIG);

    debug('registerControllers', TYPE.Controller);
    const controllerIds = registry.get(TYPE.Controller);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: interfaces.ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      const controllerMiddlewares = getMiddlewares(container, controller.constructor);

      const methodMetadatas: interfaces.ControllerMethodMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        controller.constructor
      );

      if (controllerMetadata && methodMetadatas) {
        const router = express.Router();
        const routerPath = cleanPath(`${restConfig.rootPath}${controllerMetadata.path}`);

        debug(`New route : "${routerPath}"`);

        methodMetadatas.forEach((methodMetadata: interfaces.ControllerMethodMetadata) => {
          const methodMetadataPath = cleanPath(methodMetadata.path);
          const methodMiddlewares = getMiddlewares(container, controller.constructor, methodMetadata.key);
          debug(methodMetadataPath);
          debug({ methodMiddlewares, controllerMiddlewares });
          const handler: express.RequestHandler = this.handlerFactory(
            container,
            controllerId,
            methodMetadata.key,
            controllerMetadata.json
          );
          (router as any)[methodMetadata.method](
            methodMetadataPath,
            ...controllerMiddlewares,
            ...methodMiddlewares,
            handler
          );
        });
        const app = container.get<express.Application>(APP);
        app.use(routerPath, router);
      }
    });
  }

  private handlerFactory(
    container: inversifyInterfaces.Container,
    controllerId: any,
    key: string,
    json: boolean
  ): express.RequestHandler {
    return (req: express.Request, res: express.Response, next: express.NextFunction) => {
      const result: any = container.get<any>(controllerId)[key](req, res, next);

      // try to resolve promise
      if (result && result instanceof Promise) {

        result.then((value: any) => {
          if (value !== undefined && !res.headersSent) {
            if (json) {
              res.json(value);
            } else {
              res.send(value);
            }
          }
        })
          .catch((error: any) => {
            next(error);
          });

      } else if (result !== undefined && !res.headersSent) {
        if (json) {
          res.json(result);
        } else {
          res.send(result);
        }
      }
    };
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: inversifyInterfaces.Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>((resolve) => {
        server.close(() => resolve());
      });
    } catch (e) {
    }
  }

  async start(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<interfaces.ExpressPluginConfig>(EXPRESS_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = (typeof port === 'string') ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    function onListening(): void {
      const addr = server.address();
      const bind = (typeof addr === 'string') ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }
}
