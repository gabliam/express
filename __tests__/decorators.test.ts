import { ExpressConfig, ExpressErrorConfig } from '../src/decorators';
import { ExpressConfigMetadata } from '../src/interfaces';
import { METADATA_KEY } from '../src/constants';

describe('ExpressConfig decorator', () => {
  test('should add ExpressConfig metadata to a class when decorated with @ExpressConfig', () => {
    class TestConfig {
      @ExpressConfig()
      testMethod() {}

      @ExpressConfig()
      test2Method() {}
    }

    const expressConfigMetadata: ExpressConfigMetadata[] = Reflect.getMetadata(
      METADATA_KEY.MiddlewareConfig,
      TestConfig
    );

    expect(expressConfigMetadata).toMatchSnapshot();
  });

  test('should add ExpressConfig metadata to a class when decorated with @ExpressConfig(order)', () => {
    class TestConfig {
      @ExpressConfig(50)
      testMethod() {}

      @ExpressConfig(100)
      test2Method() {}
    }

    const expressConfigMetadata: ExpressConfigMetadata[] = Reflect.getMetadata(
      METADATA_KEY.MiddlewareConfig,
      TestConfig
    );

    expect(expressConfigMetadata).toMatchSnapshot();
  });

  test('should fail when decorated multiple times the same method with @ExpressConfig', () => {
    expect(function() {
      class TestConfig {
        @ExpressConfig()
        @ExpressConfig(50)
        testMethod() {}

        @ExpressConfig(100)
        test2Method() {}
      }
      // tslint:disable-next-line:no-unused-expression
      new TestConfig();
    }).toThrowError();
  });
});

describe('ExpressErrorConfig decorator', () => {
  test('should add ExpressErrorConfig metadata to a class when decorated with @ExpressErrorConfig', () => {
    class TestConfig {
      @ExpressErrorConfig()
      testMethod() {}

      @ExpressErrorConfig()
      test2Method() {}
    }

    const expressErrorConfigMetadata: ExpressConfigMetadata[] = Reflect.getMetadata(
      METADATA_KEY.MiddlewareErrorConfig,
      TestConfig
    );

    expect(expressErrorConfigMetadata).toMatchSnapshot();
  });

  test('should add ExpressErrorConfig metadata to a class when decorated with @ExpressErrorConfig(order)', () => {
    class TestConfig {
      @ExpressErrorConfig(50)
      testMethod() {}

      @ExpressErrorConfig(100)
      test2Method() {}
    }

    const expressErrorConfigMetadata: ExpressConfigMetadata[] = Reflect.getMetadata(
      METADATA_KEY.MiddlewareErrorConfig,
      TestConfig
    );

    expect(expressErrorConfigMetadata).toMatchSnapshot();
  });

  test('should fail when decorated multiple times the same method with @ExpressErrorConfig', () => {
    expect(function() {
      class TestConfig {
        @ExpressErrorConfig()
        @ExpressErrorConfig(50)
        testMethod() {}

        @ExpressErrorConfig(100)
        test2Method() {}
      }
      // tslint:disable-next-line:no-unused-expression
      new TestConfig();
    }).toThrowError();
  });
});
