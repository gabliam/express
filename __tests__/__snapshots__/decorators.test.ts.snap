// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpressConfig decorator should add ExpressConfig metadata to a class when decorated with @ExpressConfig 1`] = `
Array [
  Object {
    "id": "TestConfig",
    "key": "testMethod",
    "order": 1,
  },
  Object {
    "id": "TestConfig",
    "key": "test2Method",
    "order": 1,
  },
]
`;

exports[`ExpressConfig decorator should add ExpressConfig metadata to a class when decorated with @ExpressConfig(order) 1`] = `
Array [
  Object {
    "id": "TestConfig",
    "key": "testMethod",
    "order": 50,
  },
  Object {
    "id": "TestConfig",
    "key": "test2Method",
    "order": 100,
  },
]
`;

exports[`ExpressErrorConfig decorator should add ExpressErrorConfig metadata to a class when decorated with @ExpressErrorConfig 1`] = `
Array [
  Object {
    "id": "TestConfig",
    "key": "testMethod",
    "order": 1,
  },
  Object {
    "id": "TestConfig",
    "key": "test2Method",
    "order": 1,
  },
]
`;

exports[`ExpressErrorConfig decorator should add ExpressErrorConfig metadata to a class when decorated with @ExpressErrorConfig(order) 1`] = `
Array [
  Object {
    "id": "TestConfig",
    "key": "testMethod",
    "order": 50,
  },
  Object {
    "id": "TestConfig",
    "key": "test2Method",
    "order": 100,
  },
]
`;
